Restructure KVStore to one library per store type

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Fixes: #13657 
Mutually depends on: https://github.com/ARMmbed/mbed-os-example-kvstore/pull/60

In #13307 we combined multiple `KVStore` types (`TDBStore`, `FileSystemStore`) and the global API into one library, `kv-global-api`. This is inconvenient for users of bare metal: if they only want one store type, they need to add dependencies for other unused types and the global API to `"requires"`. This becomes cumbersome as `FileSystem` has many dependencies.

This PR restructures KVStore to follow the same logic of [BlockDevice](https://github.com/ARMmbed/mbed-os/tree/master/storage/blockdevice): declare the interface at the top level, and treat each type/implementation as a "sub-" library. Users only need to enable the interface + the type they need. So each of the following is a library (containing `mbed_lib.json`):
* storage/kvstore: `KVStore` interface
* storage/kvstore/tdbstore: `TDBStore : KVStore` class
* storage/kvstore/filesystemstore: `FileSystemStore : KVStore` class
* storage/kvstore/securestore: `SecureStore : KVStore` class
* storage/kvstore/kvstore_global_api: optional [API to globally manage](https://os.mbed.com/docs/mbed-os/v6.4/apis/static-global-api.html) instances of KVStore, containing `kvstore_global_api` functions and `KVMap` which is a helper class for the global API
* storage/kvstore/kvconfig: configurations

Greentea and unit tests have also been moved into subdirectories of corresponding KVStore types.

~Note: The CMake support for KVStore has not been componentised yet - compilation with CMake still brings in all the unused types and dependencies, because everything is under one `mbed-storage-kvstore` target. Hopefully the CMake team will later componentise each type of KVStore into `mbed-storage-kvstore-<type>`. similar to what they've already done for `BlockDevice`s. @0xc0170~  (Update: required CMake changed included in this PR)

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

Users can enable one type of `KVStore` (e.g. `TDBStore`) without having to bring in dependencies for unused types of KVStore.

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

For a bare metal application, add `"kvstore"` and the type of KVStore to use (e.g. `"tdbstore"`) to `"requires"` in `mbed_app.json`. Note that the `KVStore` _interface_ library is `"kvstore"`, _not_ `"kv-global-api"` which now contains only the global API.
To use CMake (full profile and bare metal profile): Add the TDBStore sub-component you need (e.g. `mbed-storage-tdbstore`) and dependencies (e.g. TDBStore requires `mbed-storage-blockdevice`) to your app's `CMakeLists.txt`.

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR

Note: For the CMake support, I tested the "feature-cmake" branch of [mbed-os-example-blinky-baremetal](https://github.com/ARMmbed/mbed-os-example-blinky-baremetal), with `TDBStore store(BlockDevice::get_default_instance());` (and headers) added to `main.cpp` and the following dependencies added to the app's `CMakeLists.txt`: `mbed-storage-kvstore`, `mbed-storage-blockdevice`, `mbed-storage` (for the default block device).
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@ARMmbed/mbed-os-core @evedon @boraozgen 

----------------------------------------------------------------------------------------------------------------

LDong-Arm committed:
Turn TDBStore and FileSystemStore into standalone libraries

LDong-Arm committed:
Create a kvstore_global_api for the global API and KVMap

LDong-Arm committed:
Update include paths

LDong-Arm committed:
Update UNITTEST for kvstore

LDong-Arm committed:
Relocate kvstore unit tests into each store type's path

LDong-Arm committed:
Add separated kvstore libraries to baremetal test config

LDong-Arm committed:
Relocate kvstore Greentea tests into each store type's path

LDong-Arm committed:
KVStore: componentize CMake definitions into sub-components

LDong-Arm commented on storage/kvstore/filesystemstore/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2020 ARM Limited. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+target_include_directories(mbed-storage-kvstore
```
This should be componentised into `mbed-storage-filesystemstore` or `mbed-storage-kvstore-filesystemstore`. It's outside the scope of this task, so I'm leaving componentisation to the CMake team.

0xc0170 replied:
we would like to work with you to fix this, can you please create a story in your board to do this, we will assist.

Examples show how this is not ideally structured (for kv store I need to add like 5 libraries), we should fix dependencies.

LDong-Arm replied:
What structure would you proposed?

For BlockDevice, we have one CMake target per type of BlockDevice: https://github.com/ARMmbed/mbed-os/blob/66c05ddbf8da7873108494ec9d243c2d2ca6a283/storage/CMakeLists.txt#L7-L13

And in `mbed-storage-blockdevice`, we include them based on availability: https://github.com/ARMmbed/mbed-os/blob/66c05ddbf8da7873108494ec9d243c2d2ca6a283/storage/blockdevice/CMakeLists.txt#L4-L26

Shall we go for the same approach for KVStore?

hugueskamba replied:
@LDong-Arm ,
One of the reasons for merging the CMake work to master was so it becomes everyone's responsibility to update CMake input source files according to changes made in the file structure.
We can help you review the CMake input source files but you have to make the necessary change along with everything else.
The approach used for BlockDevice can also be used here.
Thanks.

LDong-Arm replied:
The PR in the current state already works as it is, aligning with the BlockDevice approach is a further suggestion. I can still make it here in this PR, since it's a simple change.

hugueskamba replied:
Your PR slims down what is included for baremetal.|
This is currently not the case if the `mbed-baremetal` CMake target is used with `mbed-storage-kvstore`.

LDong-Arm replied:
Done, please review again, thanks

LDong-Arm committed:
Fix astyle in KVStore unit tests

0xc0170 commented on storage/kvstore/kvstore_global_api/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+target_include_directories(mbed-storage-kvstore
+    INTERFACE
+        .
+        ./include
```
for consistency with sources below, let's remove `./` from `./xxx` in CMakelists in this PR

hugueskamba replied:
In the [PR to use relative paths](https://github.com/ARMmbed/mbed-os/commit/0909f6fac60e19e976488e5fd729160ca9e90b1a) we used `./`.

LDong-Arm replied:
It's a style convention we need to agree on. `target_sources()` below also contain relative paths but without `./` at the beginning.

LDong-Arm replied:
I've made the change, but we can revert it based on discussion.

hugueskamba replied:
No strong opinion there, I just pointed an inconsistency.

LDong-Arm replied:
Btw, I don't think we need to add the current path (`.`) as an include path - headers are inside `include`, and there's no use case for including something outside it.

LDong-Arm committed:
CMake fix: location of MBED_CONF_FILESYSTEM_PRESENT=1

LDong-Arm committed:
KVStore: drop './' from CMake paths

LDong-Arm committed:
CMake: add missing linkages of storage libraries

LDong-Arm committed:
"kvstore_global_api" requires "kvstore" for backward compatibility

PR #13908 replaces the umbrella library for all KVStore from
"kvstore_global_api" to "kvstore". Now "kvstore_global_api"
still exists but is restricted to the global API part only.

By adding "kvstore" as a dependency of "kvstore_global_api",
existing bare metal projects (e.g. mbed-bootloader) that use
the latter library name to pull in the KVStore prototype class
will continue to work.

LDong-Arm committed:
CMake fix: SFDP already moved to BlockDevice
